### ISSUE TITLE
Lambda declaration without braces fails to compile

### ISSUE THREAD

**Issue by ben-albrecht** [chapel-lang/chapel#16604]
### Summary of Problem

A lambda declaration without braces fails to compile.

It is not completely clear on the spec whether this should work, but @mppf and I consider this is a bug in the implementation.

Transferred from: https://chapel.discourse.group/t/lambda-with-statement-not-block-doesnt-compile/1086 (originally reported by @russel)

### Steps to Reproduce

**Source Code:**

Both the following function definitions compile fine:
```chapel
proc f(n: int): int { return 1; }
```
```chapel
proc f(n: int): int return 1;
```
The following lambda using a block works fine:
```chapel
var f = lambda (n: int): int { return 1; };
```
but using just a statement causes a compilation fail:
```chapel
var f = lambda (n: int): int return 1;
```

### Configuration Information

- Output of `chpl --version`: 1.23.0

**Comment by utkarsh147-del** [comment]
Please give this issue to me.

**Comment by mppf** [comment]
hi @utkarsh147-del - this issue requires some knowledge of `bison`. Generally speaking addressing this issue will involve editing `compiler/parser/chapel.ypp` and then running `make parser && make`. I don't think this issue is particularly beginner friendly but if you already have experience with `bison` grammars it might go OK. Please also see https://chapel-lang.org/contributing.html if you haven't already.

**Comment by rchinmay** [comment]
@ben-albrecht @mppf I tried to work on this issue. Here are some of the things that I found:

- Though 
```chpl 
var f = lambda (n: int): int return 1;
``` 

does not work and produces compilation error, 
```chpl
var f = lambda (n: int): int return 1;;
``` 
works well. So, I think the compilation error generated is not because the lambda expressions without braces are not supported, but it is because of the missing semi-colon. I think this syntax is fine, though usage of double-semicolon is not so frequent in any programming language (that I have used). I think that the double-semicolon is because of one for completion of return statement, and another for completion of the lambda variable declaration, and merging them to single semi-colon is not possible.
- If there is any other method of working on this other than separating the case of only using return statement (``ret_tag``) in ``lambda_decl_expr`` in ``chapel.ypp``, and specifying only one semi-colon to complete both the statements, then please tell me. I find this method would work for this case, but would produce bad or unknown behaviour for some cases. I cannot find any such problems that it could create as of now, but I feel that it is not the best way of completing two statements with a single semi-colon.
- Also, if this not expected to work, and produce error for incomplete lambda-variable declaration, or missing semi-colon, then it should be possible to do this with parser changes. I like this change, as it is safer than the previous method.

If I am wrong anywhere, or if there is any better way of such implementations, then please tell me.
Thanks, 
R. Chinmay

**Comment by mppf** [comment]
@rchinmay - that is interesting. I wonder if we would have to include `return` as a kind of expression available in a  one-expression lambda instead of as a statement.

**Comment by bradcray** [comment]
Good analysis @rchinmay!  My own intuition here would be to require the curly brackets for defining lambdas to avoid confusion.  Would that be too off-putting for people coming to Chapel from other languages with lambdas?

(FWIW, I've always found our exception regarding curly brackets and "return-only" functions to be a bit odd as well, though we've used it enough that I've always been reluctant to suggest that we should change it).

**Comment by rchinmay** [comment]
@mppf, I think including the case of return-only expression in lambda declaration in the parser to allow use of return as an expression and not a statement, would resolve this issue, where return only takes a variable, or a constant. But I don't think that would work when the value to be returned is a statement by itself, like ``return a+b;`` or ``return if a>b then a else b;`` then avoiding the semi-colon for the return expression might cause problems, since these statements inside the return block might need semi-colon for its completion. Also, it could produce some unknown bugs in the bugs, like I feel:
```chpl
proc foo() return 1;
var f = lambda (n: int) return foo();
```
could cause error, while treating return as an expression, and not a statement. If all of these work as expected by treating return as an expression, I still feel there could be cases where it could produce unknown errors. If there is some implementation of return (or any other similar keyword treated as a statement by default) as an expression and not a statement, please let me know. I will try to understand them more clearly, since I feel it could be unsafe to implement `return` as an expression, and not a statement, for this case, since I don't know if it could work for all cases.
@bradcray, I also feel it is safer to use the curly-bracket or double-semicolon in case of return-only lambda functions.

**Comment by mppf** [comment]
I think it would be possible to adjust the parser to terminate all expressions on a `;`. However I don't see that as necessarily an improvement in the language.

While the double-semicolon strategy does work today, I think I'd favor instead adjusting the parser to only allow lambdas with curly braces, as @bradcray suggested [above](https://github.com/chapel-lang/chapel/issues/16604#issuecomment-802937492). I think the double-semicolon is just odd.

**Comment by bradcray** [comment]
> I would be slightly inclined to make proc( allow single do statements.

I agree with that.  And also with the double semicolon being reasonable.
